crates/futures/src/atomics.rs

```diff
@@ -0,0 +1,326 @@
+        // being polled.
+        Notified,
+
+        // The future is blocked, waiting for something to happen. Stored here
+        // is a self-reference to the future itself so we can pull it out in
+        // `Notify` and continue polling.
+        //
+        // Note that the self-reference here is an Arc-cycle that will leak
+        // memory unless the future completes, but currently that should be ok
+        // as we'll have to stick around anyway while the future is executing!
+        //
+        // This state is removed as soon as a notification comes in, so the leak
+        // should only be "temporary"
+        Waiting(Arc<Package>),
+    }
+
+    struct Waker {
+        value: AtomicI32,
+        notified: AtomicBool,
+    };
+
+    impl Default for Waker {
+        fn default() -> Self {
+            Waker {
+                value: AtomicI32::new(0),
+                notified: AtomicBool::new(false),
+            }
+        }
+    }
+
+    impl Notify for Waker {
+        fn notify(&self, _id: usize) {
+            if !self.notified.swap(true, Ordering::SeqCst) {
+                let _ = unsafe {
+                    core::arch::wasm32::atomic_notify(
+                        &self.value as *const AtomicI32 as *mut i32,
+                        std::u32::MAX, // number of threads to notify
+                    )
+                };
+            }
+        }
+    }
+
+    fn poll_again(package: Arc<Package>) {
+        let me = match package.notified.replace(State::Notified) {
+            // we need to schedule polling to resume, so keep going
+            State::Waiting(me) => {
+                me
+            }
+
+            // we were already notified, and were just notified again;
+            // having now coalesced the notifications we return as it's
+            // still someone else's job to process this
+            State::Notified => {
+                return;
+            }
+
+            // the future was previously being polled, and we've just
+            // switched it to the "you're notified" state. We don't have
+            // access to the future as it's being polled, so the future
+            // polling process later sees this notification and will
+            // continue polling. For us, though, there's nothing else to do,
+            // so we bail out.
+            // later see
+            State::Polling => {
+                return;
+            }
+        };
+
+        // Use `Promise.then` on a resolved promise to place our execution
+        // onto the next turn of the microtask queue, enqueueing our poll
+        // operation. We don't currently poll immediately as it turns out
+        // `futures` crate adapters aren't compatible with it and it also
+        // helps avoid blowing the stack by accident.
+        let promise =
+            crate::polyfill::wait_async(&package.waker.value).expect("Should create a Promise");
+        let closure = Closure::once(Box::new(move |_| {
```

Comment:
FWIW I think [`once_into_js`](https://docs.rs/wasm-bindgen/0.2.47/wasm_bindgen/closure/struct.Closure.html#method.once_into_js) is what you want here

Reply:
I don't get it, once_into_js returns JsValue, and promise.then expects Closure. Am I missing something?

Reply:
Ah this needed a redefinition of a few functions, but I've handled that and pushed a few commits.